File: src/runtime/test_runner/expect/toHaveNthReturnedWith.rs
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};

use super::mock;
use super::{Expect, get_signature};

pub(crate) fn to_have_nth_returned_with(
    this: &Expect,
    global: &JSGlobalObject,
    frame: &CallFrame,
) -> JsResult<JSValue> {
    bun_jsc::mark_binding!();
    let [nth_arg, expected] = frame.arguments_as_array::<2>();
    let (this, returns, _value) = this.mock_prologue(
        global,
        frame.this(),
        "toHaveNthReturnedWith",
        "<green>n<r>, <green>expected<r>",
        mock::MockKind::Returns,
    )?;

    // Validate n is a number
    if !nth_arg.is_any_int() {
        return Err(global.throw_invalid_arguments(format_args!(
            "toHaveNthReturnedWith() first argument must be an integer"
        )));
    }
    let n = nth_arg.to_int32();
    if n <= 0 {
        return Err(global.throw_invalid_arguments(format_args!(
            "toHaveNthReturnedWith() n must be greater than 0"
        )));
    }

    let calls_count = u32::try_from(returns.get_length(global)?).unwrap();
    let index = u32::try_from(n - 1).unwrap(); // Convert to 0-based index

    let mut pass = false;
    let mut nth_return_value: JSValue = JSValue::UNDEFINED;
    let mut nth_call_threw = false;
    let mut nth_error_value: JSValue = JSValue::UNDEFINED;
    let nth_call_exists = index < calls_count;

    if nth_call_exists {
        let nth_result = returns.get_direct_index(global, index);
        match mock::parse_mock_result(global, nth_result)? {
            mock::MockResult::Return(value) => {
                nth_return_value = value;
                if nth_return_value.jest_deep_equals(expected, global)? {
                    pass = true;
                }
            }
            mock::MockResult::Throw(result) => {
                nth_call_threw = true;
                nth_error_value = result.get(global, "value")?.unwrap_or(JSValue::UNDEFINED);
            }
            mock::MockResult::Other => {}
        }
    }

    if pass != this.flags.get().not() {
        return Ok(JSValue::UNDEFINED);
    }

    // Handle failure
    let signature = get_signature("toHaveNthReturnedWith", "<green>n<r>, <green>expected<r>", false);

    if this.flags.get().not() {
        return mock::throw_not_failure(
            &this, global, "toHaveNthReturnedWith", "<green>n<r>, <green>expected<r>",
            format_args!("Expected mock function not to have returned on call {}", n), expected, "\nBut it did.\n",
        );
    }

    if !nth_call_exists {
        return mock::throw_nth_call_missing(&this, global, signature, calls_count, index + 1, "\n");
    }

    if nth_call_threw {
        return mock::throw_call_threw(&this, global, signature, format_args!("Call {}", n), nth_error_value);
    }

    // Diff if possible
    if expected.is_string() && nth_return_value.is_string() {
        return mock::throw_diff(&this, global, signature, format_args!("Call {}:\n", n), expected, nth_return_value);
    }

    mock::throw_expected_received(&this, global, signature, format_args!("Call {}:\n", n), expected, nth_return_value)
}

File: src/test_runner/expect/toHaveNthReturnedWith.zig
pub fn toHaveNthReturnedWith(this: *Expect, globalThis: *JSGlobalObject, callframe: *CallFrame) bun.JSError!JSValue {
    jsc.markBinding(@src());
    const thisValue = callframe.this();
    defer this.postMatch(globalThis);
    const value: JSValue = try this.getValue(globalThis, thisValue, "toHaveNthReturnedWith", "<green>n<r>, <green>expected<r>");

    const nth_arg, const expected = callframe.argumentsAsArray(2);

    // Validate n is a number
    if (!nth_arg.isAnyInt()) {
        return globalThis.throwInvalidArguments("toHaveNthReturnedWith() first argument must be an integer", .{});
    }

    const n = nth_arg.toInt32();
    if (n <= 0) {
        return globalThis.throwInvalidArguments("toHaveNthReturnedWith() n must be greater than 0", .{});
    }

    this.incrementExpectCallCounter();
    const returns = try bun.cpp.JSMockFunction__getReturns(globalThis, value);
    if (!returns.jsType().isArray()) {
        var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
        defer formatter.deinit();
        return globalThis.throw("Expected value must be a mock function: {f}", .{value.toFmt(&formatter)});
    }

    const calls_count = @as(u32, @intCast(try returns.getLength(globalThis)));
    const index = @as(u32, @intCast(n - 1)); // Convert to 0-based index

    var pass = false;
    var nth_return_value: JSValue = .js_undefined;
    var nth_call_threw = false;
    var nth_error_value: JSValue = .js_undefined;
    var nth_call_exists = false;

    if (index < calls_count) {
        nth_call_exists = true;
        const nth_result = returns.getDirectIndex(globalThis, index);
        if (nth_result.isObject()) {
            const result_type = try nth_result.get(globalThis, "type") orelse .js_undefined;
            if (result_type.isString()) {
                const type_str = try result_type.toBunString(globalThis);
                defer type_str.deref();
                if (type_str.eqlComptime("return")) {
                    nth_return_value = try nth_result.get(globalThis, "value") orelse .js_undefined;
                    if (try nth_return_value.jestDeepEquals(expected, globalThis)) {
                        pass = true;
                    }
                } else if (type_str.eqlComptime("throw")) {
                    nth_call_threw = true;
                    nth_error_value = try nth_result.get(globalThis, "value") orelse .js_undefined;
                }
            }
        }
    }

    if (pass != this.flags.not) {
        return .js_undefined;
    }

    // Handle failure
    var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
    defer formatter.deinit();

    const signature = comptime getSignature("toHaveNthReturnedWith", "<green>n<r>, <green>expected<r>", false);

    if (this.flags.not) {
        return this.throw(globalThis, comptime getSignature("toHaveNthReturnedWith", "<green>n<r>, <green>expected<r>", true), "\n\n" ++ "Expected mock function not to have returned on call {d}: <green>{f}<r>\n" ++ "But it did.\n", .{ n, expected.toFmt(&formatter) });
    }

    if (!nth_call_exists) {
        return this.throw(globalThis, signature, "\n\n" ++ "The mock function was called {d} time{s}, but call {d} was requested.\n", .{ calls_count, if (calls_count == 1) "" else "s", n });
    }

    if (nth_call_threw) {
        return this.throw(globalThis, signature, "\n\n" ++ "Call {d} threw an error: <red>{f}<r>\n", .{ n, nth_error_value.toFmt(&formatter) });
    }

    // Diff if possible
    if (expected.isString() and nth_return_value.isString()) {
        const diff_format = DiffFormatter{ .expected = expected, .received = nth_return_value, .globalThis = globalThis, .not = false };
        return this.throw(globalThis, signature, "\n\nCall {d}:\n{f}\n", .{ n, diff_format });
    }

    return this.throw(globalThis, signature, "\n\nCall {d}:\nExpected: <green>{f}<r>\nReceived: <red>{f}<r>", .{ n, expected.toFmt(&formatter), nth_return_value.toFmt(&formatter) });
}

const bun = @import("bun");
const DiffFormatter = @import("../diff_format.zig").DiffFormatter;

const jsc = bun.jsc;
const CallFrame = bun.jsc.CallFrame;
const JSGlobalObject = bun.jsc.JSGlobalObject;
const JSValue = bun.jsc.JSValue;
const mock = bun.jsc.Expect.mock;

const Expect = bun.jsc.Expect.Expect;
const getSignature = Expect.getSignature;
